File: haskell/src/Y2019/Day9.hs
module Y2019.Day9 (solve) where

import Data.List.Split (splitOn)

import Y2019.Intcode (ProgramState(..), runIntcodeXMem)

boost :: [Int] -> Int -> [Int]
boost xs n = psOutput . runIntcodeXMem 1 $ ProgramState xs [n] [] 0 0

solve :: String -> (String, String)
solve input = (part1, part2)
  where
    ns = map read $ splitOn "," input
    part1 = show $ boost ns 1
    part2 = show $ boost ns 2

File: clojure/src/y2023/day9.clj
(ns y2023.day9
  (:require
    [clojure.string :as s]))

(defn parse-line [line]
  (map read-string (re-seq #"-?\d+" line)))

(defn differences [ns]
  (last
    (reduce
      (fn [[prev acc] next] [next (conj acc (- next prev))])
      [(first ns) []]
      (rest ns))))

(defn extrapolate [history]
  (take-while
    #(not (every? (partial = 0) %))
    (iterate differences history)))

(defn find-next [history]
    (reduce
      (fn [acc xs] (+ acc (last xs)))
      0
      (reverse (extrapolate history))))

(defn solve [input]
  (let [histories (map parse-line (s/split-lines input))
        part1 (apply + (map find-next histories))
        part2 (apply + (map find-next (map reverse histories)))]
    [part1 part2]))
